dev/com.ibm.ws.microprofile.metrics.3.0/src/com/ibm/ws/microprofile/metrics30/impl/MetricRegistry30Impl.java

```diff
@@ -0,0 +1,1090 @@
+                                                           + "with a backslash `\\` ");
+                    }
+                    val = val.replace("\\,", ",");
+                    val = val.replace("\\=", "=");
+                    if (!cumulativeTags.contains(key)) {
+                        cumulativeTags.add(new Tag(key, val));
+                    }
+                }
+            }
+        } catch (NoSuchElementException e) {
+            //Continue if there is no global tags
+        }
+
+        MetricID MetricID = new MetricID(metadata.getName(), tags);
+        Class<T> metricClass = determineMetricClass(metric);
+
+        //Ensure all metrics with this name are the same type
+        validateMetricNameToSingleType(MetricID.getName(), metricClass);
+
+        /*
+         * Rest of the method officialy registers the metric
+         * Add to MetricID -> Metric Map and Name -> MetaData map
+         */
+
+        final Metric existingMetric = metricsMID.putIfAbsent(MetricID, metric);
+
+        if (existingMetric != null) {
+            throw new IllegalArgumentException("A metric named " + MetricID.getName() + " with tags " + MetricID.getTagsAsString() + " already exists");
+        }
+
+        this.metadataMID.putIfAbsent(metadata.getName(), metadataBuilder.build());
+
+        addNameToApplicationMap(MetricID);
+        return metric;
+    }
+
+    /**
+     * Adds the MetricID to an application map.
+     * This map is not a complete list of metrics owned by an application,
+     * produced metrics are managed in the MetricsExtension
+     *
+     * @param name
+     */
+    protected void addNameToApplicationMap(MetricID metricID) {
+        String appName = getApplicationName();
+        addNameToApplicationMap(metricID, appName);
+    }
+
+    /**
+     * Adds the MetricID to an application map given the application name.
+     * This map is not a complete list of metrics owned by an application,
+     * produced metrics are managed in the MetricsExtension
+     *
+     * @param metricID metric ID of metric that was added
+     * @param appName  applicationName
+     */
+    public void addNameToApplicationMap(MetricID metricID, String appName) {
+        // If it is a base metric, the name will be null
+        if (appName == null)
+            return;
+        ConcurrentLinkedQueue<MetricID> list = applicationMap.get(appName);
+        if (list == null) {
+            ConcurrentLinkedQueue<MetricID> newList = new ConcurrentLinkedQueue<MetricID>();
+            list = applicationMap.putIfAbsent(appName, newList);
+            if (list == null)
+                list = newList;
+        }
+        list.add(metricID);
+    }
+
+    public void unRegisterApplicationMetrics() {
+        unRegisterApplicationMetrics(getApplicationName());
+    }
+
+    public void unRegisterApplicationMetrics(String appName) {
+        ConcurrentLinkedQueue<MetricID> list = applicationMap.remove(appName);
```

Comment:
You might have a NPE here as `appName` could be `null`.

Reply:
Fixed in https://github.com/OpenLiberty/open-liberty/pull/12299/commits/9edbbeb2b39fae1a4d57c5f9a9c65443ee9ffbc9